include/openmc/shared_array.h

```diff
@@ -0,0 +1,119 @@
+#ifndef OPENMC_SHARED_ARRAY_H
+#define OPENMC_SHARED_ARRAY_H
+
+//! \file shared_array.h
+//! \brief Shared array data structure
+
+#include<memory>
+
+
+namespace openmc {
+
+//==============================================================================
+// Class declarations
+//==============================================================================
+
+// The SharedArray is an array that is capable of being appended to in an
+// thread safe manner by use of atomics. It only provides protection for the
+// use cases currently present in OpenMC. Namely, it covers the scenario where
+// multiple threads are appending to an array, but no threads are reading from
+// or operating on it in any other way at the same time. Multiple threads can
+// call the thread_safe_append() function concurrently and store data to the
+// object at the index returned from thread_safe_append() safely, but no other
+// operations are protected.
+template <typename T> 
+class SharedArray { 
+
+private: 
```

Comment:
We follow C++ core guidelines, one of which is
[NL.16](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#nl16-use-a-conventional-class-member-declaration-order): Use `public` before `protected` before `private` order